geo/geo.go

```diff
@@ -0,0 +1,243 @@
+			return true
+		}
+	}
+	return false
+}
+
+func (f *IPFilter) AllowIP(ip string) (bool, error) {
+	if f.Empty() {
+		return true, nil
+	}
+
+	var loc *Location
+	if f.NeedGeoInfo() {
+		loc = f.GetLocation(ip)
+	} else {
+		loc = &Location{IP: ip}
+	}
+
+	return f.AllowLocation(loc), nil
+}
+
+func (f *IPFilter) GetLocation(ip string) *Location {
+	for _, p := range f.Providers {
+		if p.Ready() {
+			loc := getLocationFromProvider(ip, p.(GeoProvider))
+			if !loc.Empty() {
+				return &loc
+			}
+		}
+	}
+	return &Location{CountryCode: "UNKNOWN", IP: ip}
+}
+
+func (f *IPFilter) AllowLocation(loc *Location) bool {
+	if loc.Empty() {
+		return true
+	}
+
+	for _, l := range f.Disallow {
+		if l.Match(loc) {
+			return false
+		}
+	}
+
+	empty := true
+	for _, l := range f.Allow {
+		if l.Match(loc) {
+			return true
+		}
+		if !l.Empty() {
+			empty = false
+		}
+	}
+
+	return empty
+}
+
+func getLocationFromProvider(ip string, p GeoProvider) Location {
+	loc, err := p.GetLocation(ip)
+	if err != nil {
+		log.Println(err)
+	}
+	return *loc
+}
+
+func GetJsonFile(url, filename string) ([]byte, error) {
+	if s, err := os.Stat(filename); os.IsNotExist(err) || s.Size() == 0 {
+		f, err := os.Create(filename)
+		if err != nil {
+			return nil, err
+		}
+		defer f.Close()
+
+		client := http.Client{
+			Timeout: 5 * time.Second,
+		}
+		resp, err := client.Get(url)
+		if err != nil {
+			return nil, err
+		}
+		defer resp.Body.Close()
+		b, err := ioutil.ReadAll(resp.Body)
+		if !json.Valid(b) {
+			return nil, errors.New("invalid json")
+		}
+		_, err = f.Write(b)
+		if err != nil {
+			os.Remove(filename)
+			return nil, err
+		}
+	}
+	f, err := os.Open(filename)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	b, err := ioutil.ReadAll(f)
+	if err != nil {
+		return nil, err
+	}
+	return b, nil
+}
+
+func (f *IPFilter) AddProvider() {
+	if len(f.Providers) == 0 {
+		if f.Download {
+			aws := NewAWSProvider(f.DBPath)
+			aws.MaybeUpdate()
+			gcp := NewGCPProvider(f.DBPath)
+			gcp.MaybeUpdate()
+			mm := NewMaxMindProvider(f.DBPath)
+			mm.MaybeUpdate()
+			f.Providers = append(f.Providers, aws, gcp, mm)
+		}
+
+		ip2c := NewIP2CProvider()
+		ip2c.MaybeUpdate()
+		f.Providers = append(f.Providers, ip2c)
+	}
+}
+
+func (f *IPFilter) UpdateDataFile() {
+	for {
+		time.Sleep(1 * time.Hour)
+		for idx, p := range f.Providers {
+			if len(p.FileName()) == 0 {
+				continue
+			}
+			filename := p.FileName()
+			_, err := os.Stat(filename)
+			if !p.NeedUpdate() && err == nil {
+				continue
+			}
+
+			tmpFile, err := ioutil.TempFile("", filename+"-*")
+			if err != nil {
+				log.Print(err)
+			}
+
+			p.SetFileName(tmpFile.Name())
+			err = p.MaybeUpdate()
+			if err != nil {
+				log.Print(err)
+				continue
+			}
+			err = os.Rename(tmpFile.Name(), filename)
+			if err != nil {
+				log.Print(err)
+				os.Remove(tmpFile.Name())
+				continue
+			}
+			p.SetFileName(filename)
+			f.Providers[idx] = p
```

Comment:
It's better to use `Providers []*GeoProvider` instead of `Providers []GeoProvider` to avoid this

Reply:
I tested before, cause `range` will passing value not pointer.

Reply:
But now all provider is already pointer, so I think it should work. You can test again now

Reply:
In this case i need to change the value of elems inside slice,but elem `p` here is the copy of the value, not a pointer of value.

Reply:
`p` is already the pointer when you add it in `AddProvider`, see this example: https://play.golang.org/p/-QlpEpMznc_e